lms/events/subscriber.py

```diff
@@ -0,0 +1,29 @@
+from pyramid.events import subscriber
+
+from lms.models.file import File
+
+
+class FilesDiscoveredEvent:
+    """Notify that a file has been found in an LMS."""
+
+    def __init__(self, request, values):
+        """Initialize the event.
+
+        :param request: Pyramid request object
+        :param values: List of dicts of file information
+        """
+        self.request = request
+        self.values = values
+
+
+@subscriber(FilesDiscoveredEvent)
+def files_discovered(event):
+    """Record discovered files in the DB."""
+
+    application_instance_id = (
+        event.request.find_service(name="application_instance").get().id
+    )
+    for value in event.values:
+        value["application_instance_id"] = application_instance_id
+
+    File.bulk_upsert(session=event.request.db, values=event.values)
```

Comment:
You wouldn't need to pass in the session, but you'd still need a request to find the service in the first place, so I feel like it's a bit of a wash. Long story short you need the request around either way, and I'm not sure that's much different? Which one is simpler?

```python
event.request.find_service(name="bulk_upsert").upsert(models.File, values=event.values).
# vs.
File.bulk_upsert(session=event.request.db, values=event.values)
```

### Are we concentrating on too few solutions?

If your dog is stuck in the banister and you're trying to workout if a land-mine or a tambourine is the correct solution, the answer might be neither.

I think I can see what you mean about injecting the DB session. It feels like a different kind of method from say `ApplicationInstance.update_lms_data()`, as that method modifies the model in the abstract. The model remains completely unconcerned about the existence of DBs.

This feels more like `db_session.add()` or something. You can't do it without a session, and it only accepts a model and some data. Is this something could tack onto the db session where ever we make it?

### I don't think you can really separate the bulk-upsertiness(tm) from the model

You can obviously move the code outside of the model as we do here with a mixin, or with a service but I don't think you can actually decouple the fact you need co-operation from the model itself when it comes to updating. You might be able to change what you update meaningfully (although I doubt you would) but I don't think you can vary the indexing requirements. It's pretty much the correct answer or broken.

So we could put the code that does the upserting separate from the variables in the model that are required to make it work, and it _appears_ like it _could_ work on anything, but in truth crashes on anything other than the specific models that cater for it. Similarly the models have some random config attached to them that _looks_ like it doesn't do anything but is actually _implicitly_ coupled with the service, rather than more explicitly coupled with the mixin.

So my feeling is a service in this case pretends it's decoupled, whilst actually just concealing the real coupling. Whereas a `BulkUpsertMixin` actively advertises that this model supports that action. Try calling `bulk_upsert` on a model that doesn't support it and the method isn't even there.

Maybe we could split the difference and have `BulkUpsertMixin` that only adds the config, and have something like this in the service?

`if not isinstance(BulkUpsertMixin)...`

### It would help me out if you could nail down some terms a bit for me

I think there might be some slop in my understanding of  how you're using a few of the terms and phrases here.

When you say harmonious with our general approach is this more of a "where we'd like to go" rather "where we are/what we've done" type of thing? Our approach as it appears in the code is clearly pretty varied, so I guess you mean "our future approach" or something instead? For example this is pretty much a slimmed down rip of the same thing from Checkmate: https://github.com/hypothesis/checkmate/blob/main/checkmate/models/db/mixins.py#L44.
Having methods on the models to create / update is also we do in `LtiLaunches.add()`, `GroupInfo.upsert_instructor()`. Lord knows that's up in `h`. 

Normally when people say put the logic in services, it's short hand for put the _**business logic**_ in services, rather than all logic. Seeing as the suggested `BulkUpsert` service wants to be model agnostic, it seems like it can't contain any business type logic. It seems a pure DB concern, which typically would be dealt with in the DB layer. Are you saying here we don't want any kind of logic in our DB layer at all regardless of the type of logic or is there a conflation here?

I wonder if you could also go into more detail about what you mean by thin models in this context. I feel like there's an implication I don't get in "having to inherit a lot of methods". It kind of reads to me like this is an effort for the model classes to load all these methods, or for us as programmers to add them all one by one, rather than something that just happens automatically through inheritance if you add things to either the BASE (as we do in `BaseClass.update_from_dict()`) or in the `BulkUpsertMixin`?

I think there's also a confusion for me about the model become "fat" by virtue of using a mixin. In this case is the "fatness" the fact that the the model has some functionality? Not the fact the model has a lot of actual code inside the class? As both the mixin and the service keep the actual implementation outside of the specific model.

Reply:
I've had a crack at this and changed the base session object to be a class of our own from which we can add methods like this. I've kept the mixin, but it only adds the required fields. I think this achieves most of what we want:

* It keeps the code in the DB layer (which I think is distinct from the model layer?)
* It keeps the models thin, as long as you don't consider the DB layer to be the model layer
* It provides a way of adding DB functionality which can be shared everywhere (although models still have to have the right fields)
* The mixin still allows us to advertise which models do, and do not support this

What's been lost:

* The link is less obvious as the mixin class doesn't directly relate to `bulk_upsert`
* It does at least give us a location for documentation though, so we can have a pointer to the session object, which makes this discoverable at least
* You can obviously still call this on things that don't support it

Reply:
What's the reason for adding methods to a `Session` subclass, rather than using a service? Do we need to invent a new thing here? Or would a service do the job just fine?

```python
class BulkUpsertService:
    def __init__(self, db):
        self._db = db

    def bulk_upsert(self, model_class, values):
        ...

def factory(_context, request):
    return BulkUpsertService(request.db)
```

And then:

```python
request.find_service(name="bulk_upsert").bulk_upsert(models.File, event.values)
```

I guess I'm not really seeing the downside to this? And therefore I'm not seeing why we'd introduce a new solution.

Services seem more scalable as you can keep adding more of them, whereas the `Session` class will get too big if you add too many methods to it, and you could end up having to find some way of breaking it up.

If you really don't want to use a service then I think the model mixin that you had originally might be a better approach than the DB session-based one. I think it's simpler to understand. Have a custom SQLAlchemy `Session` subclass seems a bit clever to me, and you still end up needing a mixin anyway.

Reply:
There's an ADR about the services layer here: https://github.com/hypothesis/h/blob/master/docs/arch/adr-002.rst The Nando Farestan blog post that it links to is very good. 

I don't think we really make a distinction between "business logic" and "DB logic", we just put both in services. I don't really think of us as having a "DB layer" as such, technically there is a `db/` directory but it just contains some necessary SQLAlchemy boiler. Do we have much reason to distinguish between "DB logic" and "business logic"?

There are some slide about thin models in this presentation, starting on slide 50: https://docs.google.com/presentation/d/1O-rl1qr0diNNAwACLAe7StLeZEi4kuIvhfwrX83AMKM/edit#slide=id.g4c957d27dd_0_566 Basically the reasons why we move SQL queries up into the services layer is that multi-model queries don't make sense on any one particular model class; because model objects tend to get all over the whole codebase (and not mocked in tests) so we want them to be close to "value objects" so they don't end up tightly coupling a lot of code; services scale better because you can have many services for one model (whereas if all those methods were on the model class it'd be one too-big class).

Reply:
> Maybe we could split the difference and have `BulkUpsertMixin` that only adds the config, and have something like this in the service?
> 
>     if not isinstance(BulkUpsertMixin)...

I think it's more Pythonic to avoid the `isinstance()` and just assume that the class has the necessary attrs on it (and go ahead and crash if it doesn't). Just use duck typing, the code will be simpler. You don't need the mixin.

Reply:
I've written a long thing below, but I'd suggest you don't read it. We're at the point where text tennis probably isn't going to give us a good understanding, or be a good use of our time. I'd like to discuss this in person, because I don't think I'm going to be able to understand it without being able to ask questions interactively.

----

I don't think we make any distinction between pure DB functionality and business logic type things, but I'm sort of suggesting here that that's a problem rather and potentially a case of not having gotten into it rather than the solution to the problem.

I think services for us are under specified and sort of a golden hammer type approach. Any code which doesn't fit into one of our predefined buckets ends up lumped into services for lack of any other likely locations to put it.

Seeing as services are basically any code, and any code can be written as a service (including a view if you're mad) then asking if something _can_ be done as a service seems to me to be arguing from the wrong direction. _Everything_ could be a service, but what _should_ be a service? What does making it a service add over it being normal code in another location? What's special about services that isn't true of a plain class? etc.

The reasons to make distinctions between different layers and responsibilities in your services and other parts of the app are largely the same ones as motivate having services in the first place again:

* To have different areas responsible for different things
* To make it easier to understand
* To make it convenient to call

I think it's beneficial for the services to have an identity and set of expectations for what's in there (or not) for the same reason it's a good idea to do that everywhere else.

### I think you said why services are good, but not why this doesn't apply to other things

You've given some reasons you don't want a bunch of specific SQL in all the model objects and I totally agree, but what I think is missing from that is how this is a counter argument to any of the other suggested solutions. Basically you state here that services do these things, and it's implied that other things don't, and I think I need to understand why to internalise this.

The benefits of SQL code in services and out of models you say:

### Multi-model queries don't make sense on any one particular model class 

The session solution fixes this directly. Another way to say this is also that multi-model queries make as much sense on any model as long as they aren't defined there, which the mixin would seem to fix.

### We want them to be close to "value objects" so they don't end up tightly coupling a lot of code

I think I need some explanation on this one. I think this might be a mocking concern? It might be more awkward to mock out a single method on the value objects. That definitely seems like a mark against the mixins. On the session object this seems like a direct reason to do it: we don't mock out basic DB access like adding, deleting and querying. We're just adding one more to that list. It might be a massive error and counter practice to mock out the `BulkInsertService` and the `BulkInsertService` alone because actually it's acting more like a DB functionality. 

We'd make that error because we don't differentiate between lower level DB functionality (which we should not mock) and higher level stuff (which we should).

### You can have many services for one model (whereas if all those methods were on the model class it'd be one too-big class).

The session solution directly addresses this.

With the models I asked above, but I don't think you addressed it. Does a model become "larger" in your mind when it exposes more functionality, even if that functionality is defined in one place and re-used? As in if the code in the file is exactly the same size on the disk, but there's different inheritance which results in more or less methods on the class, is that model "bigger" now?

I tend to define "too-big" as too big to look at when you open code in an editor on the whole. I get it's possible to add so many methods that you can end up with name spacing difficulties or auto complete becomes a giant mess, but I don't think we're talking about getting that far are we? Is this "conceptually" too big, vs. lines of code too big?